### DISCUSSION TITLE
How to use bytecode to create contract instance in Certora and Foundry?

### DISCUSSION THREAD

**Discussion by dt6120** [Cyfrin/assembly-evm-opcodes-and-formal-verification-course#24] [Q&A] [answered]
Hi everyone, I am trying to solve the section 2 challenge. In the challenge contract, it says to find a counter example to prove that function `exp2` in below contracts are not equivalent. Challenge says helper contract was not intended to be verified, so I am not taking reference of the verified code.

Challenge: https://sepolia.etherscan.io/address/0x3DbBF2F9AcFB9Aac8E0b31563dd75a2D69148D64#code
Helper: https://sepolia.etherscan.io/address/0xf2dde8E99f583b5354441EB7a141042419082596#code

One approach is to fuzz both of them on forked Sepolia rpc with the same input and revert on different outputs.

Now I want to use `Halmos` and `Certora` to formally verify this. Is there a way to create a contract instance using `bytecode` in Certora and Foundry (for Halmos)? I am thinking of copying Challenge contract code and using Helper contract bytecode, and then passing them to the tools to havoc and find counter example.

**Comment by EngrPips** [comment]
I haven't tried to solve this challenge yet, but talking about deploying a contract by its bytecode you look into Foundry `vm.etch` cheatcode https://book.getfoundry.sh/cheatcodes/etch. Maybe that will help.

**Reply by dt6120** [reply]
Thanks for suggesting the cheatcode. It works as intended with the fuzz test, the Certora prover is getting the following error because cheatcode requires to import Test.sol:
```
CVL
The type `StdStorage` will not be accessible in CVL code
Reason(s):
struct field `finds` cannot be expressed in CVL: [mapping types are not supported]
```

 I also tried using the `create` opcode in assembly like below:
 ```solidity
contract Challenge {

    IChallenge h;

    constructor() {
        // copied the deployed bytecode of the Helper contract from Etherscan
        bytes memory bytecode = hex"<CONTRACT_CREATION_BYTECODE + RUNTIME_BYTECODE>";
        uint256 size = bytecode.length;
        address addr;
        assembly {
            // get the free memory pointer
            let fmp := mload(0x40)
            // store the bytecode in the memory 
            mstore(fmp, bytecode)
            // create address with value `0` by fetching bytecode from offset `fmp` and of size `size`
            addr := create(0, fmp, size)

            // revert if unable to create address
            if iszero(addr) {
                revert(0, 0)
            }
        }
        h = IChallenge(addr);
    }

    function exp2(uint256 x) private pure returns (uint256 result) {
        // challenge code copy-pasted here
    }

    // test case will be assert(compare(x))
    // counter example will be input which returns false
    function compare(uint256 x) external returns (bool) {
        return exp2(x) == h.exp2(x);
    }
}
 ```

Running the test case on `compare` function always results in `EvmError: Revert` as `h.exp2(x)` reverts. I am not able to figure out why.

**Reply by EngrPips** [reply]
That can be tough. I didn't bother learning Certora as I prefer Halmos because of it's simplicity. Maybe you can copy the bytecode of the `h` contract to `evm playground` and call the `exp2` function on it with the appropriate argument and step into the process of the `exp2` function opcode by opcode to see what is actually breaking.

**Reply by dt6120** [reply]
The bytecode has the function signature of `exp2(uint256)`. It seems the helper contract is not being is created with the bytecode intended:
```solidity
// after assembly block to create contract
console.logBytes(addr.code);
// returns 0x
```

**Reply by dt6120** [reply, marked as answer]
Figured it out. When creating new bytes variable, `bytes memory code = hex"";`, Solidity first stores the length of bytes in the free memory pointer (`mload(0x40)`) and then starts storing the value from the next 32 bytes which changes the actual bytecode offset to (`add(mload(0x40), 0x20)`). Below code worked:

```solidity
bytes memory code = hex"";
address addr;
assembly {
    let size := mload(0x80)
    addr := create(0, add(0x80, 0x20), size)
}
h = IChallenge(addr);
console.log(h.exp2(1));
```

Working with Halmos also now, thanks @EngrPips !

**Reply by EngrPips** [reply]
I am glad you figured it out, and I must confess that I appreciate you sharing the solution here. Yeah, Halmos makes my life easier lol, I just want to be a good security researcher, not a formal verification savant. To me, Certora is too much stress lol, but I know many people will object to this.

**Reply by dt6120** [reply]
Haha, Certora indeed has a very steep learning curve.

**Reply by EngrPips** [reply]
Indeed I didn't come here to learn Formal Verification but instead to be a good Security researcher.
